File: forumClient/src/ApiResponses/ApiSuccessResponse.ts
import { ApiResponse, ApiResponseType, ApiResponseWithMetadata } from './ApiResponse';
import type { AllowedUserActions } from '@/Actions/AllowedUserActions';

export class ApiSuccessResponse<D> extends ApiResponse<D> {
    constructor(
        message: string,
        data: D | undefined,
        public allowedUserActions?: AllowedUserActions,
        statusCode?: number
    ) {
        super(true, message, data, ApiResponseType.Success, undefined, statusCode);
        this.allowedUserActions = allowedUserActions;
    }
}

export class ApiSuccessResponseWithMetadata<D, M> extends ApiResponseWithMetadata<D, M> {
    constructor(
        message: string,
        data: D | undefined,
        metadata: M,
        public allowedUserActions?: AllowedUserActions,
        statusCode?: number
    ) {
        super(true, message, data, metadata, ApiResponseType.Success, undefined, statusCode);
        this.allowedUserActions = allowedUserActions;
    }
}

export class ApiSuccessResponses {
    static withData<D>(
        message: string,
        data: D,
        allowedUserActions?: AllowedUserActions,
        statusCode?: number
    ): ApiSuccessResponse<D> {
        return new ApiSuccessResponse(message, data, allowedUserActions, statusCode);
    }

    static withDataAndMetadata<D, M>(
        message: string,
        data: D,
        metadata: M,
        allowedUserActions?: AllowedUserActions,
        statusCode?: number
    ): ApiSuccessResponseWithMetadata<D, M> {
        return new ApiSuccessResponseWithMetadata(message, data, metadata, allowedUserActions, statusCode);
    }

    static withoutData(
        message: string,
        statusCode?: number
    ): ApiSuccessResponse<unknown> {
        return new ApiSuccessResponse(message, null, undefined, statusCode);
    }
}

export interface PaginatedData<D, S = unknown> {
    rows: D;
    pageNumber: number;
    rowsPerPage: number;
    totalPages: number;
    summary?: S;
}

File: server/RestApiServer.Endpoints/ApiResponses/ApiSuccessResponse.cs
using RestApiServer.Endpoints.Security;

namespace RestApiServer.Endpoints.ApiResponses
{
    public class ApiSuccessResponse<D> : ApiResponse<D>
    {
        public Dictionary<string, List<AllowedUserAction>>? AllowedUserActions { get; set; }

        public ApiSuccessResponse(
            string message, 
            D? data, 
            Dictionary<string, List<AllowedUserAction>>? allowedUserActions = null,
            int? statusCode = null)
            : base(true, message, data, ApiResponseType.Success, null, statusCode)
        {
            AllowedUserActions = allowedUserActions;
        }
    }

    public class ApiSuccessResponseWithMetadata<D, M> : ApiResponseWithMetadata<D, M>
    {
        public Dictionary<string, List<AllowedUserAction>>? AllowedUserActions { get; set; }

        public ApiSuccessResponseWithMetadata(
            string message, 
            D? data, 
            M metadata,
            Dictionary<string, List<AllowedUserAction>>? allowedUserActions = null,
            int? statusCode = null)
            : base(true, message, data, metadata, ApiResponseType.Success, null, statusCode)
        {
            AllowedUserActions = allowedUserActions;
        }
    }

    public static class ApiSuccessResponses
    {
        public static ApiSuccessResponse<D> WithData<D>(
            string message, 
            D data, 
            Dictionary<string, List<AllowedUserAction>>? allowedUserActions = null,
            int? statusCode = null) =>
            new(message, data, allowedUserActions, statusCode);

        public static ApiSuccessResponseWithMetadata<D, M> WithData<D, M>(
            string message, 
            D data, 
            M metadata,
            Dictionary<string, List<AllowedUserAction>>? allowedUserActions = null,
            int? statusCode = null) =>
            new(message, data, metadata, allowedUserActions, statusCode);

        public static ApiSuccessResponse<object> WithoutData(
            string message,
            int? statusCode = null) =>
            new(message, null, null, statusCode);
    }

    // Pagination support
    public class PaginatedData<D, S>
    {
        public required D Rows { get; set; }
        public required int PageNumber { get; set; }
        public required int RowsPerPage { get; set; }
        public required int TotalPages { get; set; }
        public S? Summary { get; set; } // Optional summary
    }
}
